### PR TITLE
docs: component migration - general aspects

### PR DESCRIPTION
**Pull request by bajtos**
This is a part of documentation for the spike #4099 How to migrate LB3 components.

There is quite a lot of techniques and content to cover, I am getting overwhelmed by the amount of thinking and writing needed. To make things more manageable, I decided to split the work into smaller chunks.

In this pull request, I am starting to create basic structure for the migration guide and adding a draft content for the following migration aspects:
- [x] Migrate component infrastructure
- [x] Migrate current context
- [x] Migrate mixins

This is a part of a spike, so I should not be changing the real docs, but then I figured out that since I have already wrote the content, it will give more value to our users to put the content up, even if it's not finished yet. My intention is to land this pull request even if the content is not in the shape would you eventually want and create follow-up stories to make any substantial improvements as required by reviewers. (This the the part where I am getting back to our usual spike process - do the research and then open follow-up stories to implement the proposed changes.)

@raymondfeng @jannyHou @emonddr: please review

**Out of scope**

(will be covered by follow-up pull requests)

- Migrate Models, Entities and Repositories
- Migrate REST API
- Migrate API transports
- Authentication & authorization
- Migrate Introspection

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- New tests added or existing tests modified to cover all changes
- Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- Affected artifact templates in `packages/cli` were updated
- Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: docs/site/migration/components/project-layout.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+As part of this task, we should also revisit the pages "Creating components"
+and "Using components", to make sure they are consistent with this migration
+guide and the updated generator (project template). For example, the pages
+should explain how to receive and provide component config via `@config` and
+`this.configure()` API, and so on.
+
+When the issue is created, then add a "note" paragraph here with a link to the
+new issue and a short explanation that users must make those changes manually
+for now.
+-->
```
Comment (bajtos) [inline docs/site/migration/components/project-layout.md:131]: I'll replace this FIXME comment with a note pointing to the GitHub issue I'll create after we agree this is the direction to take.

File: docs/site/migration/components/current-context.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+   continuation-local-storage to provide static per-request storage that can be
+   accessed from anywhere inside a LoopBack application (an Express middleware,
+   a model method, a connector, etc.).
+
+In LoopBack 4, extensions should use `@inject` decorators to access contextual
+information. For example:
+
+- `@inject(key)` and `@inject.getter(key)` to receive values from the context
+- `@inject.setter(key)` to obtain a setter function for writing values to the
+  context
```
Comment (raymondfeng) [inline docs/site/migration/components/current-context.md:40]: We should mention that `TRANSIENT` scope should be used for such bindings in order to receive `per-request` contextual data.

File: docs/site/migration/components/overview.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  - Extensions for Extension Points defined by 3rd-party component
+- Any other values to be bound in `Application`'s context:
+  - Classes
+  - Providers
+  - Arbitrary bindings (instances of `Binding` class)
+
+As the last resort, LoopBack 4 components can also modify the target application
+directly by calling `Application` APIs (this is similar to LoopBack 3 approach).
+
+To make the migration guide easier to navigate, we split component-related
```
Comment (raymondfeng) [inline docs/site/migration/components/overview.md:70]: We should mention https://github.com/strongloop/loopback-next/pull/5304 and give an example to show how to use an application as a component to allow `boot` artifacts from a sub-application.

File: docs/site/migration/components/project-layout.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+This allows the component to receive configuration. App developers can provide
+environment-specific configuration by using `component-config.{env}.json` files.
+
+## LoopBack 4 layout
+
+As explained in [Creating components](../../Creating-Components.md) and
+[Using components](../../Using-components.md), a typical LoopBack 4 component is
+an npm package exporting a Component class.
+
```
Comment (raymondfeng) [inline docs/site/migration/components/project-layout.md:53]: We should recommend `lb4 extension` command to scaffold a package to expose components.

File: docs/site/migration/components/project-layout.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+component project in a series of incremental changes done within the same
+repository. We recommend to create a new project using
+[`lb4 extension`](../../Extension-generator.md) CLI command and then
+incrementally migrate artifacts from the original LoopBack 3 component to the
+new project.
+
+<!--
+FIXME: create a follow-up task to update Extension template to follow
+the latest style as used by `extensions/metrics`:
+- inject the target app and component config to the component class,
```
Comment (raymondfeng) [inline docs/site/migration/components/project-layout.md:118]: It's also possible to build a component booter to load declarative json files similar as LB3. I plan to contribute a `Configuration` booter to load `json/js/yaml/rc` files from `configs` directory and use `app.configure` to bind such entries.

File: docs/site/migration/components/project-layout.md (hunk 1 above)
Comment (bajtos) [reply docs/site/migration/components/project-layout.md:131]: Created https://github.com/strongloop/loopback-next/issues/5336

File: docs/site/migration/components/overview.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+This is a placeholder page, the task of adding content is tracked by the
+following GitHub issue:
+[loopback-next#3955](https://github.com/strongloop/loopback-next/issues/3955)
+" %}
+
+<!--
+FIXME: Create a new follow-up task to create a diagram per Raymond's comment:
+> Maybe a diagram would help if it shows the handshake between an application
+> and a component as well as typical artifacts exported from a component.
+-->
```
Comment (bajtos) [inline docs/site/migration/components/overview.md:24]: Created: https://github.com/strongloop/loopback-next/issues/5337

File: docs/site/migration/components/overview.md (hunk 3 above)
Comment (bajtos) [reply docs/site/migration/components/overview.md:70]: I am afraid I don't see how the booter for loading artifacts from a sub-application is related to migrating LB3 extensions to LB4? Can you @raymondfeng please open a follow-up issue to describe in more detail what are you looking for?

File: docs/site/migration/components/project-layout.md (hunk 4 above)
Comment (bajtos) [reply docs/site/migration/components/project-layout.md:53]: I am already recommending `lb4 extension` elsewhere, see https://github.com/strongloop/loopback-next/pull/5262/files#diff-d72cf51faca1367bfb582f0995256991R111

File: docs/site/migration/components/project-layout.md (hunk 5 above)
Comment (bajtos) [reply docs/site/migration/components/project-layout.md:118]: Sure. Such booter is out of scope of my pull request though. First, it's not implemented yet. Second, this FIXME comment is just a placeholder, the changes are tracked by the following GitHub issue I created few moments ago: https://github.com/strongloop/loopback-next/issues/5336

**Comment by bajtos** [comment]
Thank you @raymondfeng @jannyHou @emonddr for the review. I addressed the remaining bits in 4f1effc and will proceed to land this PR shortly.

**Comment by bajtos** [comment]
Follow-up stories:
- #5336
- #5337

**Pull request merged**